### PR TITLE
Make save function more universal to accept any number of 1D or 2D node or edge features

### PR DESCRIPTION
**Pull request by maxiimilian**
## Describe your changes

This pull request contains three changes -- one major and two minor changes.

The major change modifies how edge and node features are converted from `pyg.data.Data` objects into `torch.Tensor`s before saving. If we assume that all node/edge features (features in the following) are numeric, a conversion to float32 will always work. Further, my proposed change assumes that the features can be either 1D (vector of edge lengths) or 2D (node position, vdiff vector between source and target nodes). Based on this, I propose a universal function that reshapes any 1D vector into a 2D column vector, which allows the concatenation of any number of features in a loop. 

Using the new function, we also make the saving of `len` as edge feature explicit by including it in the default `edge_features` list. 

Two minor bugs are fixed:
- Default arguments of functions should not be mutables (list, dicts, any object). See https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments.
- The `hide_ticks` argument for `nx.draw` seems to be deprecated because it caused an error with my networkx v 3.2

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [x] I have updated the documentation to cover introduced code changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging)

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: src/weather_model_graphs/save.py [hunk 1]
```diff
@@ -59,6 +61,13 @@ def to_pyg(
             "install weather-mode-graphs[pytorch] to enable writing to torch files"
         )
 
+    # Default values for arguments
+    if edge_features is None:
+        edge_features = ["len", "vdiff"]
```
Comment (joeloskarsson) [inline src/weather_model_graphs/save.py:66]: This makes sense as a default as it matches neural-lam (https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/create_mesh.py#L84) :+1:

File: src/weather_model_graphs/visualise/plot_2d.py [hunk 2]
```diff
@@ -13,7 +13,7 @@ def nx_draw_with_pos(g, with_labels=False, **kwargs):
     if ax is None:
         _, ax = plt.subplots(figsize=(10, 10))
     networkx.draw_networkx(
-        ax=ax, G=g, pos=pos, hide_ticks=False, with_labels=with_labels, **kwargs
+        ax=ax, G=g, pos=pos, with_labels=with_labels, **kwargs
```
Comment (joeloskarsson) [inline src/weather_model_graphs/visualise/plot_2d.py:16]: `hide_ticks` seems to be new from networkx version 3.3 (https://networkx.org/documentation/stable/reference/generated/networkx.drawing.nx_pylab.draw_networkx.html). Looks like this is the version we require https://github.com/mllam/weather-model-graphs/blob/2456064543200af1adfd17db0d1ccbb5f7441ca8/pyproject.toml#L11
so if you upgrade to 3.3 then this change should not be necessary.

File: src/weather_model_graphs/save.py [hunk 3]
```diff
@@ -77,16 +86,20 @@ def to_pyg(
...
-            raise NotImplementedError(node_features_values)
-        return pyg_g.pos.to(torch.float32)
+    def _concat_pyg_features(pyg_g: "pyg.data.Data", features: List[str]) -> torch.Tensor:
+        """Convert features from pyg.Data object to torch.Tensor.
+        Each feature should be column in the resulting 2D tensor (n_features, n_edges or n_nodes).
+        Note, this function can handle node AND edge features.
+        """
+        v_concat = []
+        for f in features:
+            v = torch.tensor(pyg_g[f])  # make sure we have torch tensor
```
Comment (joeloskarsson) [inline src/weather_model_graphs/save.py:96]: Just for my understanding, why is this necessary? Do you have a specific case where `pyg_g[f]` is not tensor already?

File: src/weather_model_graphs/save.py [hunk 4]
```diff
@@ -77,16 +86,20 @@ def to_pyg(
...
-        # TODO: handle features of different types more generally, i.e. both single ("len") values and tuples (like "vdiff")
-        return torch.cat((pyg_g.len.unsqueeze(1), pyg_g.vdiff), dim=1).to(torch.float32)
-
-    def _get_node_features(pyg_g):
-        if node_features != ["pos"]:
-            raise NotImplementedError(node_features_values)
-        return pyg_g.pos.to(torch.float32)
+    def _concat_pyg_features(pyg_g: "pyg.data.Data", features: List[str]) -> torch.Tensor:
+        """Convert features from pyg.Data object to torch.Tensor.
+        Each feature should be column in the resulting 2D tensor (n_features, n_edges or n_nodes).
```
Comment (joeloskarsson) [inline src/weather_model_graphs/save.py:91]:
```suggestion
        Each feature should be column in the resulting 2D tensor (n_edges or n_nodes, n_features).
```

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
This is great, thanks a lot for this contribution! :smile: 

I went over the code and left some small comments. I don't think there is any need for adding additional tests here as 
1. the saving functionality with default feature arguments is already tested in https://github.com/mllam/weather-model-graphs/blob/2456064543200af1adfd17db0d1ccbb5f7441ca8/tests/test_save.py#L18 
2. we don't have any good functionality to add additional edge/node features that is now saveable with this more universal approach and could be used in a test (as far as I remember, let me know if I am wrong)

There are a couple small additional things that would be good to do before merging this:

1. The linting seems to fail due to the code formatting. This should be easily fixed by just running the pre-commit hooks once. 
2. Go ahead and add an entry to the changelog! :slightly_smiling_face:

File: src/weather_model_graphs/visualise/plot_2d.py (hunk 2 above)
Comment (maxiimilian) [reply src/weather_model_graphs/visualise/plot_2d.py:16]: Oh, I should have checked that. I will revert the change.

File: src/weather_model_graphs/save.py (hunk 3 above)
Comment (maxiimilian) [reply src/weather_model_graphs/save.py:96]: No, good point. The data from `pyg.data.Data` should always be tensors. I will remove it.

**Review by joeloskarsson — APPROVED** [review]
Everything good now, I'll merge this in. Thanks again for this contribution!
